File: scripts/sample02_02.m
%[text] # Sample 2-2
%[text] ## 画像データの入出力
%[text] 動画像処理 
%[text] 画像処理特論
%[text] 村松 正吾 
%[text] 動作確認: MATLAB R2023a
%[text] ## Input and output of images
%[text] Video processing
%[text] Advanced Topics in Image Processing
%[text] Shogo MURAMATSU
%[text] Verified: MATLAB R2023a
%%
%[text] ### サンプル画像の準備
%[text] (Preparation of sample image)
%[text] 本サンプルで利用する画像データを収めたdata フォルダにパスをとおして，サンプル動画を準備。
%[text] Create a path to the data folder that contains images used in this sample, and prepare a sample video.
addpath('./data')
close
% Preparation of a sample video
mkCalcioAvi
%[text] ### 動画像サンプルの再生
%[text] (Video sample play)
implay('calcio.avi');
%%
%[text] ### 入力動画の準備
%[text] (Preparation of input video)
%[text] VideoReaderオブジェクトの生成
%[text] Instantiation of VideoReader object
videoReader = VideoReader('calcio.avi');
frameRate   = videoReader.FrameRate;
%%
%[text] ### 出力動画の準備
%[text] (Preparation of output video)
%[text] VideoWriterオブジェクトの生成
%[text] 非圧縮で保存
%[text] Instantiation of VideoWriter object
%[text] Save without any compression.
videoWriter  = VideoWriter('calcio_gray.avi','Uncompressed AVI');
videoWriter.FrameRate = frameRate;
%%
%[text] ### フレーム毎の処理
%[text] (Frame-by-frame processing)
%[text] フレームごとのグレースケール変換
%[text] Frame-by-frame grayscale conversion
videoWriter.open();
while(videoReader.hasFrame())
    % Reading frame
    pictureIn = videoReader.readFrame();
    % Frame processing
    pictureOut = rgb2gray(pictureIn);
    % Writing frame
    videoWriter.writeVideo(pictureOut);
end
videoWriter.close();
%%
%[text] ### 動画像処理結果の再生
%[text] (Output video play)
implay('calcio_gray.avi');
%%
%[text] © Copyright, Shogo MURAMATSU, All rights reserved.

%[appendix]{"version":"1.0"}
%---
%[metadata:view]
%   data: {"layout":"inline","rightPanelPercent":40}
%---

File: python/sample02_02.py
# -*- coding: utf-8 -*-
#%% Sample 2-2
#%% 画像データの入出力
# 動画像処理 
# 
# 画像処理特論
# 
# 村松 正吾 
# 
# 動作確認: Python 3.7, PyTorch 1.8
#%% Input and output of images
# Video processing
# 
# Advanced Topics in Image Processing
# 
# Shogo MURAMATSU
# 
# Verified: Python 3.7, PyTorch 1.8
# サンプル画像の準備
# (Preparation of sample image)
# 
# 本サンプルで利用する画像データを収めたdata フォルダにパスをとおして，サンプル動画を準備。
# 
# Create a path to the data folder that contains images used in this sample, 
# and prepare a sample video.
#import requests 
import torch
import torchvision
import torchvision.io 

# Preparation of a sample video
#mkCalcioAvi

# 入力動画の準備
# (Preparation of input video)
# 
# VideoReaderオブジェクトの生成
# 
# Instantiation of VideoReader object

framesIn,_,info = torchvision.io.read_video('./data/pycalcio.avi',0, pts_unit='sec')
frameRate = info['video_fps']
nFrames = framesIn.size(0)

# フレーム毎の処理
# (Frame-by-frame processing)
# 
# フレームごとのグレースケール変換
# 
# Frame-by-frame grayscale conversion

#%% 
rgb2gray = torchvision.transforms.Grayscale(num_output_channels=3)
framesIn = framesIn.permute(0,3,1,2)
framesOut = torch.empty([])
for iFrame in range(nFrames):
    # Read frame
    pictureIn = framesIn[iFrame]
    # Process frame
    pictureOut = rgb2gray(pictureIn)
    print(pictureOut.size())    
    # Write frame
    if iFrame == 0:
        framesOut = pictureOut.unsqueeze(0)
    else:
        framesOut = torch.cat((framesOut,pictureOut.unsqueeze(0)),0) 

#
framesOut = framesOut.permute(0,2,3,1)    
torchvision.io.write_video('./tmp/pycalcio_gray.avi', framesOut, fps=frameRate, video_codec='libx264')
#%% 
# © Copyright, Shogo MURAMATSU, All rights reserved.